waku/v2/discv5/discover.go

```diff
@@ -34,23 +33,20 @@ type PeerConnector interface {
 }
 
 type DiscoveryV5 struct {
-	params      *discV5Parameters
-	host        host.Host
-	config      discover.Config
-	udpAddr     *net.UDPAddr
-	listener    *discover.UDPv5
-	localnode   *enode.LocalNode
-	metrics     Metrics
-	peerChannel *peerChannel
+	params    *discV5Parameters
+	host      host.Host
+	config    discover.Config
+	udpAddr   *net.UDPAddr
+	listener  *discover.UDPv5
+	localnode *enode.LocalNode
+	metrics   Metrics
 
 	peerConnector PeerConnector
 	NAT           nat.Interface
 
 	log *zap.Logger
 
-	started atomic.Bool
-	cancel  context.CancelFunc
-	wg      *sync.WaitGroup
+	*protocol.CommonService[peermanager.PeerData]
```

Comment:
Rather than updating the commonService to include this channel and data which is specific to discovery protocols only. 
Wouldn't be better to use composition and reuse this PeerChannel across all discovery services?
Here it feels like we are trying to fit into existing CommonService and it is not looking clean.

Reply:
I understood the reasoning behind this, but maybe we can do it in a better way.
Passing an empty struct in start of a service which doesn't need channel is not looking neat.
Also, PeerChannel is a common pattern in all discovery services to communicate with peerConnector...so i was thinking maybe it is better to extract that out of discovery and reuse it. 
One approach could be ,we can use options pattern to indicate to CommonService whether peerChannel is required or not and use/enable it accordingly in CommonService. 
There would be no change for existing other services, only for services which require working with peerChannel can set this option with a value and CommonService can internally handle it. This way code will look cleaner.

Reply:
I thought of this approach as well, but why do we want to put generic data channel when we know it is peerChannel. Instead of this, i think it might be better to use options pattern and make channel optional, which are used only by services that need it.
How about something like below?

`type CommonService struct {
    ...
    chan PeerChannel
}

CommonService.WithPeerChannel(c Peer.Data){
  ...
}

func (c *CommonService)Start(){
     ... //other starting tasks 
     //channel specific tasks...
}

`

Reply:
Hmm, make sense. 
Did not like the style of CommonService and then CommonChanService. You could maybe embed generic channel itself with options pattern? 
If not possible, then proceed with 2 service types.

Reply:
Was just thinking about this and realized that instead of using generics, can't you use an interface type for the channel? 
Then you can use options pattern to achieve this with CommonService itself.

Reply:
I like the idea of having separate services (i.e. `CommonService` and `CommonChanService`) with the idea of not affecting existing protocols, since they don't need to use this channel at all. But looking at this PR, something on which I'm not entirely sold is why use generics, if the purpose of this channel is just to pass a `peermanager.PeerData`, which is used only for passing information to the peer connector.

I'd suggest to besides having two separate base objects: `CommonService` and `CommonChanService`, to also rename the latter to `CommonDiscoveryService`, and use peerstore.PeerData directly. There's probably no need for having generics in this as this component will probably be used only by discovery services.

Reply:
> I'd suggest to besides having two separate base objects: CommonService and CommonChanService, to also rename the latter to CommonDiscoveryService, and use peerstore.PeerData directly. There's probably no need for having generics in this as this component will probably be used only by discovery services.

This is something i am also fine with. But i think @harsh-98 wanted to include PeerConnector service as well. Which is why we cannot use the term CommonDiscoveryService and rather probably can be done using options pattern WithPeerChannel to enable channel.